client/go/outline/vpn/nmconn_linux.go

```diff
@@ -0,0 +1,188 @@
+	for retries := 20; retries > 0; retries-- {
+		slog.Debug(nmLogPfx+"trying to find TUN device ...", "tun", c.nmOpts.TUNName)
+		dev, err = c.nm.GetDeviceByIpIface(c.nmOpts.TUNName)
+		if dev != nil && err == nil {
+			return
+		}
+		slog.Debug(nmLogPfx+"waiting for TUN device to be available", "err", err)
+		time.Sleep(50 * time.Millisecond)
+	}
+	return nil, errSetupVPN(nmLogPfx, "failed to find TUN device", err, "tun", c.nmOpts.TUNName)
+}
+
+func configureCommonProps(props map[string]map[string]interface{}, opts *nmConnectionOptions) {
+	props["connection"] = map[string]interface{}{
+		"id":             opts.Name,
+		"interface-name": opts.TUNName,
+	}
+}
+
+func configureTUNProps(props map[string]map[string]interface{}) {
+	props["tun"] = map[string]interface{}{
+		// The operating mode of the virtual device.
+		// Allowed values are 1 (tun) to create a layer 3 device and 2 (tap) to create an Ethernet-like layer 2 one.
+		"mode": uint32(1),
+	}
+}
+
+func configureIPv4Props(props map[string]map[string]interface{}, opts *nmConnectionOptions) {
+	props["ipv4"] = map[string]interface{}{
+		"method": "manual",
+
+		// Array of IPv4 addresses. Each address dictionary contains at least 'address' and 'prefix' entries,
+		// containing the IP address as a string, and the prefix length as a uint32.
+		"address-data": []map[string]interface{}{{
+			"address": opts.TUNAddr4.String(),
+			"prefix":  uint32(32),
+		}},
+
+		// A lower value has a higher priority.
+		// Negative values will exclude other configurations with a greater value.
+		"dns-priority": -99,
+
+		// routing domain to exclude all other DNS resolvers
+		// https://manpages.ubuntu.com/manpages/jammy/man5/resolved.conf.5.html
+		"dns-search": []string{"~."},
+
+		// NetworkManager will add these routing entries:
+		//   - default via 10.0.85.5 dev outline-tun0 table 13579 proto static metric 450
+		//   - 10.0.85.5 dev outline-tun0 table 13579 proto static scope link metric 450
+		"route-data": []map[string]interface{}{{
+			"dest":     "0.0.0.0",
+			"prefix":   uint32(0),
+			"next-hop": opts.TUNAddr4.String(),
+			"table":    opts.RoutingTable,
+		}},
+
+		// Array of dictionaries for routing rules. Each routing rule supports the following options:
+		// action (y), dport-end (q), dport-start (q), family (i), from (s), from-len (y), fwmark (u), fwmask (u),
+		// iifname (s), invert (b), ipproto (s), oifname (s), priority (u), sport-end (q), sport-start (q),
+		// supress-prefixlength (i), table (u), to (s), tos (y), to-len (y), range-end (u), range-start (u).
+		//
+		//   - not fwmark "0x711E" table "113" priority "456"
+		"routing-rules": []map[string]interface{}{{
+			"family":   unix.AF_INET,
+			"priority": opts.RoutingPriority,
+			"fwmark":   opts.FWMark,
+			"fwmask":   uint32(0xFFFFFFFF),
+			"invert":   true,
+			"table":    opts.RoutingTable,
+		}},
+	}
+
+	dnsList := make([]uint32, 0, len(opts.DNSServers4))
+	for _, dns := range opts.DNSServers4 {
```

Comment:
I don't think you need to restrict the IP version of the DNS servers. An IPv4 network can use IPv6 dns servers, the same way IPv6 networks use IPv4 DNS servers.

Reply:
But this is not how NetworkManager works. Here is what NetworkManager accepts for `ipv4.dns`:

>  dns | array of uint32 | Array of IP addresses of DNS servers (as network-byte-order integers)

If you want to set IPv6 DNS, you need to use `ipv6.dns`:

> dns | array of byte array | Array of IP addresses of DNS servers (in network byte order)

I can add `DNSServers6` to this PR, through we haven't enabled IPv6 for the connection yet (because we haven't provide any IPv6 addresses).